src/problems/MetalAdvectionProblem/test_sne.cpp

```diff
@@ -0,0 +1,545 @@
+	    [=] AMREX_GPU_DEVICE(int i, int j, int k, amrex::Array4<const Real> const &state) noexcept {
+		    Real const rho = state(i, j, k, HydroSystem<NewProblem>::density_index);
+		    return (rho);
+	    },
+	    dir);
+
+	proj["scalar"] = computePlaneProjection<amrex::ReduceOpSum>(
+	    [=] AMREX_GPU_DEVICE(int i, int j, int k, amrex::Array4<const Real> const &state) noexcept {
+		    Real const rhoZ = state(i, j, k, Physics_Indices<NewProblem>::pscalarFirstIndex);
+		    return (rhoZ);
+	    },
+	    dir);
+	return proj;
+}
+
+// Implement User-defined diode BC
+template <>
+AMREX_GPU_DEVICE AMREX_FORCE_INLINE void AMRSimulation<NewProblem>::setCustomBoundaryConditions(const amrex::IntVect &iv, amrex::Array4<Real> const &consVar,
+												int /*dcomp*/, int /*numcomp*/, amrex::GeometryData const &geom,
+												const Real /*time*/, const amrex::BCRec * /*bcr*/,
+												int /*bcomp*/, int /*orig_comp*/)
+{
+	auto [i, j, k] = iv.dim3();
+	amrex::Box const &box = geom.Domain();
+	const auto &domain_lo = box.loVect3d();
+	const auto &domain_hi = box.hiVect3d();
+	const int klo = domain_lo[2];
+	const int khi = domain_hi[2];
+	int kedge = 0;
+	int normal = 0;
+
+	if (k < klo) {
+		kedge = klo;
+		normal = -1;
+	} else if (k > khi) {
+		kedge = khi;
+		normal = 1.0;
+	}
+
+	const double rho_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::density_index);
+	const double x1Mom_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::x1Momentum_index);
+	const double x2Mom_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::x2Momentum_index);
+	double x3Mom_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::x3Momentum_index);
+	const double etot_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::energy_index);
+	const double eint_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::internalEnergy_index);
+
+	if ((x3Mom_edge * normal) < 0) { // gas is inflowing
+		x3Mom_edge = -1. * consVar(i, j, kedge, HydroSystem<NewProblem>::x3Momentum_index);
+	}
+
+	consVar(i, j, k, HydroSystem<NewProblem>::density_index) = rho_edge;
+	consVar(i, j, k, HydroSystem<NewProblem>::x1Momentum_index) = x1Mom_edge;
+	consVar(i, j, k, HydroSystem<NewProblem>::x2Momentum_index) = x2Mom_edge;
+	consVar(i, j, k, HydroSystem<NewProblem>::x3Momentum_index) = x3Mom_edge;
+	consVar(i, j, k, HydroSystem<NewProblem>::energy_index) = etot_edge;
+	consVar(i, j, k, HydroSystem<NewProblem>::internalEnergy_index) = eint_edge;
+}
+
+auto problem_main() -> int
+{
+
+	const int ncomp_cc = Physics_Indices<NewProblem>::nvarTotal_cc;
+	amrex::Vector<amrex::BCRec> BCs_cc(ncomp_cc);
+
+	for (int n = 0; n < ncomp_cc; ++n) {
+		for (int i = 0; i < AMREX_SPACEDIM; ++i) {
+			// diode boundary conditions
+			if (i == 2) {
+				BCs_cc[n].setLo(i, amrex::BCType::ext_dir);
+				BCs_cc[n].setHi(i, amrex::BCType::ext_dir);
+			} else {
+				BCs_cc[n].setLo(i, amrex::BCType::int_dir); // periodic
+				BCs_cc[n].setHi(i, amrex::BCType::int_dir); // periodic
+			}
+		}
+	}
+
```

Comment:
Add this here:
```
	// set random state
	const int seed = 42;
	amrex::InitRandom(seed, 1); // all ranks should produce the same values
```
Currently, I think different MPI ranks may produce different random numbers, which will cause the supernova rate to be wrong.

Reply:
Why would the SN-rate be wrong? I am not sure I follow completely.

Reply:
The default random number generator seed is set to the MPI rank + 1:
https://github.com/AMReX-Codes/amrex/blob/6f039ae18e892c606c06fab7111bf11b57c8975a/Src/Base/AMReX.cpp#L643

So different SNe events will be generated on each MPI rank, which will lead to inconsistent/wrong results.

Reply:
The correct way to do it in demonstrated in the RandomBlast problem here: https://github.com/quokka-astro/quokka/blob/fe6215e0ebbf4ae1d3d75125c005a750f8a691da/src/problems/RandomBlast/blast.cpp#L352

Reply:
But the rate of SN isn't set by the random seed, right?

Reply:
The problem is there will be missing SNe. If each rank determines independently where the SNe go off, but only sets off SNe on its local grids, then there are SNe that are 'missing' -- they never go off, because other grids generate their own events, but only check *their* own SN list.

Reply:
I'm not sure whether there's a fortuitous cancellation of errors here or not. But it should not be coded this way, and it is necessary to reason very carefully about the algorithm in order to determine whether this caused the SN rate to be wrong in your simulations.

Reply:
You could check the output logs to verify that the actual SN rate was what you input.

Reply:
For reproducibility, I think it would be still better to use the same RNG seed for all the processes.

Reply:
The random seed is now same.